clang/lib/CIR/CodeGen/LoweringPrepare.cpp

```diff
@@ -865,30 +893,379 @@ void LoweringPreparePass::lowerThreeWayCmpOp(CmpThreeWayOp op) {
+
+  clang::CIRGen::CIRGenModule &cgm = *this->cgm;
+  clang::CIRGen::CIRGenCXXABI &cxxABI = cgm.getCXXABI();
+  clang::CIRGen::CIRGenBuilderTy &builder = cgm.getBuilder();
+
+  builder.setInsertionPointAfter(getGlobalOp);
+  Block *getGlobalOpBlock = builder.getInsertionBlock();
+  // TODO(CIR): This is too simple at the moment. This is only tested on a
+  // simple test case with only the static local var decl and thus we only have
+  // the return. For less trivial examples we'll have to handle shuffling the
+  // contents of this block more carefully.
+  Operation *ret = getGlobalOpBlock->getTerminator();
+  ret->remove();
+  builder.setInsertionPointAfter(getGlobalOp);
+
+  clang::CIRGen::CIRGenFunction cgf{cgm, builder, true};
+  cgf.CurFn = getGlobalOp->getParentOfType<cir::FuncOp>();
+  cgm.setCurrentCIRGenFn(&cgf);
+
+  // Inline variables that weren't instantiated from variable templates have
+  // partially-ordered initialization within their translation unit.
+  bool nonTemplateInline =
+      varDecl.isInline() &&
+      !isTemplateInstantiation(varDecl.getTemplateSpecializationKind());
+
+  // We only need to use thread-safe statics for local non-TLS variables and
+  // inline variables; other global initialization is always single-threaded
+  // or (through lazy dynamic loading in multiple threads) unsequenced.
+  bool threadsafe = astCtx->getLangOpts().ThreadsafeStatics &&
+                    (varDecl.isLocalVarDecl() || nonTemplateInline) &&
+                    !varDecl.getTLSKind();
+
+  // If we have a global variable with internal linkage and thread-safe
+  // statics are disabled, we can just let the guard variable be of type i8.
+  bool useInt8GuardVariable = !threadsafe && globalOp.hasInternalLinkage();
+
+  cir::IntType guardTy;
+  clang::CharUnits guardAlignment;
+  if (useInt8GuardVariable) {
+    guardTy = cir::IntType::get(&getContext(), 8, /*isSigned=*/true);
+    guardAlignment = clang::CharUnits::One();
+  } else {
+    // Guard variables are 64 bits in the generic ABI and size width on ARM
+    // (i.e. 32-bit on AArch32, 64-bit on AArch64).
+    if (::cir::MissingFeatures::useARMGuardVarABI()) {
+      llvm_unreachable("NYI");
+    } else {
+      guardTy = cir::IntType::get(&getContext(), 64, /*isSigned=*/true);
+      ::cir::CIRDataLayout layout{globalOp->getParentOfType<mlir::ModuleOp>()};
+      guardAlignment =
+          clang::CharUnits::fromQuantity(layout.getABITypeAlign(guardTy));
+    }
+  }
+  auto guardPtrTy = cir::PointerType::get(guardTy);
+
+  // Create the guard variable if we don't already have it (as we might if
+  // we're double-emitting this function body).
+  cir::GlobalOp guard = getStaticLocalDeclGuardAddress(varDecl);
+  if (!guard) {
+    // Mangle the name for the guard.
+    SmallString<256> guardName;
+    {
+      llvm::raw_svector_ostream out(guardName);
+      cxxABI.getMangleContext().mangleStaticGuardVariable(varDecl.getRawDecl(),
```

Comment:
If the global is static_local, can we add more attributes to early pass in the name of the guard during CIRGen, and just use that here. Have you explored that?